libraries/libmesosphere/source/board/nintendo/nx/kern_k_device_page_table.cpp

```diff
@@ -619,24 +619,417 @@ namespace ams::kern::board::nintendo::nx {
+        /* Clear the output size. */
+        *out_mapped_size = 0;
+
+        /* Get the size, and validate the address. */
+        const u64 size = pg.GetNumPages() * PageSize;
+        MESOSPHERE_ASSERT((device_address & ~DeviceVirtualAddressMask) == 0);
+        MESOSPHERE_ASSERT(((device_address + size - 1) & ~DeviceVirtualAddressMask) == 0);
+
+        /* Ensure that the region we're mapping to is free. */
+        R_UNLESS(this->IsFree(device_address, size), svc::ResultInvalidCurrentMemory());
+
+        /* Ensure that if we fail, we unmap anything we mapped. */
+        auto unmap_guard = SCOPE_GUARD { this->UnmapImpl(device_address, size, true); };
+
+        /* Iterate, mapping device pages. */
+        KDeviceVirtualAddress cur_addr = device_address;
+        for (auto it = pg.begin(); it != pg.end(); ++it) {
+            /* Require that we be able to map the device page. */
+            R_UNLESS(IsHeapVirtualAddress(it->GetAddress()), svc::ResultInvalidCurrentMemory());
+
+            /* Get the physical address for the page. */
+            const KPhysicalAddress phys_addr = GetHeapPhysicalAddress(it->GetAddress());
+
+            /* Map the device page. */
+            const u64 block_size = it->GetSize();
+            size_t mapped_size = 0;
+            R_TRY(this->MapDevicePage(std::addressof(mapped_size), num_pt, max_pt, phys_addr, block_size, cur_addr, device_perm));
+
+            /* Advance. */
+            cur_addr         += block_size;
+            *out_mapped_size += mapped_size;
+
+            /* If we didn't map as much as we wanted, break. */
+            if (mapped_size < block_size) {
+                break;
+            }
+        }
+
+        /* We're done, so cancel our guard. */
+        unmap_guard.Cancel();
+
+        return ResultSuccess();
     }
 
-    bool KDevicePageTable::IsFree(KDeviceVirtualAddress address, u64 size) const {
-        MESOSPHERE_UNIMPLEMENTED();
+    void KDevicePageTable::UnmapImpl(KDeviceVirtualAddress address, u64 size, bool force) {
+        MESOSPHERE_ASSERT((address & ~DeviceVirtualAddressMask) == 0);
+        MESOSPHERE_ASSERT(((address + size - 1) & ~DeviceVirtualAddressMask) == 0);
+
+        /* Get the memory manager and page table manager. */
+        KMemoryManager &mm     = Kernel::GetMemoryManager();
+        KPageTableManager &ptm = Kernel::GetPageTableManager();
+
+        /* Make a page group for the pages we're closing. */
+        KPageGroup pg(std::addressof(Kernel::GetBlockInfoManager()));
+
+        /* Walk the directory. */
+        u64 remaining = size;
+        while (remaining > 0) {
+            const size_t l0_index = (address / DeviceRegionSize);
+            const size_t l1_index = (address % DeviceRegionSize)    / DeviceLargePageSize;
+            const size_t l2_index = (address % DeviceLargePageSize) / DevicePageSize;
```

Comment:
This directory-walking code is duplicated a few times. Perhaps write an iterator for it and deduplicate it?